### PR TITLE
{WIP} (MODULES-6861) Types use resource_api

### PR DESCRIPTION
**Pull request by tphoney**

### CODE REVIEW

File: lib/puppet_x/puppetlabs/netdev_stdlib/check.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require 'puppet_x'
+module PuppetX::NetdevStdlib
+  # Helper function to check if we should load the resource_api
+  class Check
+    def self.use_old_netdev_type
+      use_old_netdev = false
+      # is this an agent running on a network device, there are only a few of them
+      use_old_netdev = %w[aristaeos ios_xr nexus].include? Facter.value('operatingsystem').downcase
```
Comment (shermdog) [inline lib/puppet_x/puppetlabs/netdev_stdlib/check.rb]: Verified these are correct via puppet-agent 5.5.1

File: lib/puppet/type/domain_name.rb [hunk 2]
```diff
@@ -1,18 +1,40 @@
...
-      else fail "value #{value.inspect} is invalid, must be a String."
+      validate do |value|
+        if value.is_a? String then super(value)
+        else raise "value #{value.inspect} is invalid, must be a String."
+        end
       end
     end
   end
+else
+  binding.pry
```
Comment (shermdog) [inline lib/puppet/type/domain_name.rb]: Crufty pry.  breaks tests.

File: lib/puppet/type/port_channel.rb [hunk 3]
```diff
@@ -1,84 +1,145 @@
...
   end
+else
+  require 'puppet/resource_api'
+
+  Puppet::ResourceApi.register_type(
+    name: 'port_channel',
+    docs: 'Network Device Link Aggregation Group',
+    features: ['remote_resource'],
+    attributes: {
+      ensurable:      {
```
Comment (shermdog) [inline lib/puppet/type/port_channel.rb]: This should be

```ruby
      ensure:      {
        type:    'Enum[present, absent]',
        desc:    'Whether this network interface should be enabled on the target system.',
        default: 'absent'
      },
```

**Comment by shermdog** [comment]
There's also some loading issues with `puppet describe`  I wasn't able to get to the root of it though.

> [root@localhost type]# puppet describe --list --trace
Error: Could not autoload puppet/type/tacacs_server_group: cannot load such file -- puppet/resource_api
/opt/puppetlabs/puppet/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/opt/puppetlabs/puppet/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/etc/puppetlabs/code/environments/production/modules/netdev_stdlib/lib/puppet/type/tacacs_server_group.rb:38:in `<top (required)>'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:68:in `load'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:68:in `load_file'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:83:in `block in loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:81:in `each'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:81:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:208:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/metatype/manager.rb:53:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:59:in `initialize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:243:in `new'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:243:in `main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:383:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:672:in `exit_on_fail'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:137:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:73:in `execute'
/opt/puppetlabs/puppet/bin/puppet:5:in `<main>'
Error: Could not run: Could not autoload puppet/type/tacacs_server_group: cannot load such file -- puppet/resource_api
/opt/puppetlabs/puppet/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/opt/puppetlabs/puppet/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/etc/puppetlabs/code/environments/production/modules/netdev_stdlib/lib/puppet/type/tacacs_server_group.rb:38:in `<top (required)>'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:68:in `load'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:68:in `load_file'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:83:in `block in loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:81:in `each'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:81:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/autoload.rb:208:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/metatype/manager.rb:53:in `loadall'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:59:in `initialize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:243:in `new'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/describe.rb:243:in `main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:383:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:672:in `exit_on_fail'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:137:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:73:in `execute'
/opt/puppetlabs/puppet/bin/puppet:5:in `<main>'

**Comment by tphoney** [comment]
@shermdog spoke to @DavidS about the describe, we should talk about this later. I fixed the other issues

**Comment by DavidS** [comment]
I've tried that quickly with a test module and puppet 5.5.0, and `puppet describe --list` worked fine:

```
david@davids:~/git/tmp/blubb$ bundle exec -- puppet describe --modulepath spec/fixtures/modules/ blubb

blubb
=====



Parameters
----------

- **blubb**
    the hot sht. (a Boolean)
    Valid values are `true` (also called `true, true`), `false` (also called
    `false, false`). 

- **ensure**
    Whether this resource should be present or absent on the target system.
    (a Enum[present, absent])
Valid values are `absent`, `present`. 

- **flubb**
    the cold sht. (a Integer)
Values can match `/^-?\d+$/`.

- **name**
    The name of the resource you want to manage. (a String)
    Values can match `//`.
david@davids:~/git/tmp/blubb$ bundle exec -- puppet --version
5.5.0
david@davids:~/git/tmp/blubb$ 
```

The exception is complaining about `cannot load such file -- puppet/resource_api`. Do you have that in your puppet install?

**Comment by shermdog** [comment]
Superseded by #30

**Pull request closed without merging**